### ISSUE TITLE
Release thread for OpenMM 8.0

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#3610]
In the latest dev build, the errors on Mac described in #3495 seem to have disappear.  We don't know why, just as we don't know why they appeared in the first place.  But given that, I suggest we move forward with building a beta of the next release.  It contains changes needed for ML.

Are there any final features we want to get in before building the beta?  Don't worry about bug fixes and documentation changes.  Those can go in after the beta.  I'm just asking about new features.

Do we want to call it 7.8 or 8.0?

**Comment by peastman** [comment]
Does the deafening silence mean everyone agrees and I should start working on building a beta?

**Comment by raimis** [comment]
Yes, let's make a beta, so it is easier to check the NNP functionalities. For now, let's name 7.8.

**Comment by peastman** [comment]
@jchodera @giadefa @tmarkland any comments?  Which version number do you prefer?

**Comment by tmarkland** [comment]
I think we decided on 8.0 for this one. Proceeding with the beta sounds great to me.

**Comment by jchodera** [comment]
Apologies for the delay in responding---I've been traveling.

The most important functionality to test is to make sure we can make the whole `openmm-ml` ecosystem conda-installable in a single line to deploy accelerated QML. We haven't been able to test that without a dev release. Perhaps a `dev` release of that would be much more useful since we are typically reluctant to build more than one `beta`?

From the "OpenMM 8 status and priorities" document, here are the desiderata I laid out Feb 22:

* It would be good for everyone to quickly look through all open issues marked as [bugs](https://github.com/openmm/openmm/issues?q=is%3Aissue+is%3Aopen+label%3Abug+) and [enhancements](https://github.com/openmm/openmm/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement+) and make sure we didn’t miss important things
* We didn’t address fast long-range correction computation (critical for alchemical free energy calculations) for CustomNonbondedForce in 7.7—how much work would it be to address that now?
https://github.com/openmm/openmm/issues/3229 
* Now that we are using BAOAB by default (LangevinMiddleIntegrator), we can probably achieve significant practical speedups with higher default proton masses (1.5 amu) and (https://github.com/openmm/openmm/issues/3227) including some auto-tuning tools that we recommend by default:
https://github.com/openmm/openmm/issues/3403 
* Finish up automated benchmarking to make it easier for us to produce and use benchmark data
https://github.com/openmm/openmm/pull/3386 
* Work with @egallicc to finish up generalizing ATMMetaForce for alchemical free energy calculations and making it conda-installable?
https://github.com/Gallicchio-Lab/openmm-atmmetaforce-plugin
* Experimental AMD HIP/ROCm platform plugin (via https://github.com/openmm/openmm/pull/3338 )
* This should also be a good opportunity to do some major cleanup of things that aren’t as tidy:
  * Clean up and add structure to examples directory:
https://github.com/openmm/openmm/issues/730
All the files have been dumped into one place right now, which is supremely unhelpful to new users trying to make sense of >30 files
  * Update makefiles for examples:
https://github.com/openmm/openmm/issues/3355
  * Address units modernization and interoperability (e.g. with OpenFF):
Can we add support for OpenFF untis natively / pint?
Would need a quick call with OpenFF devs.

From this list, I still have to finish up the [automated benchmarks PR](https://github.com/openmm/openmm/pull/3386) which I can do quickly so it will be easier for folks to run benchmarks and share them with us.

Thoughts on the others?

**Comment by egallicc** [comment]
Thank you @jchodera and all. I have a [draft conda-forge recipe](https://github.com/conda-forge/staged-recipes/pull/19081) for the [ATMMetaForce plugin](https://github.com/Gallicchio-Lab/openmm-atmmetaforce-plugin) awaiting review.  Yes, the plugin is rather specific to traditional bonded/non-bonded MM force fields and could likely benefit from some level of generalization. Also documentation of the API, unit testing, etc.

**Comment by peastman** [comment]
> We didn’t address fast long-range correction computation (critical for alchemical free energy calculations) for CustomNonbondedForce in 7.7

There are several changes that improve it: #3520, #3552, and #3606.

> including some auto-tuning tools that we recommend by default

This would be a nice feature for someone to work on.  Right now I think it's still at the research stage.  Someone needs to experiments with various ways of tuning and see what works well.

> Finish up automated benchmarking to make it easier for us to produce and use benchmark data

If you want to finish it up, that would be great.  Since it only affects the benchmark script, I think it's fine if that doesn't happen until after the beta goes out.

> Experimental AMD HIP/ROCm platform plugin

It's still very experimental.

> Clean up and add structure to examples directory:
> Update makefiles for examples:

This would be nice, and can also be done post-beta.

> Can we add support for OpenFF untis natively / pint?

I'm not convinced that's a good idea.  It's certainly something that would need a lot of design and discussion, so not happening in this release.

**Comment by peastman** [comment]
@jchodera given the above comments, are you ok with releasing a beta now?

It would be nice to have coordinated versions of OpenMM, OpenMM-Torch, and OpenMM-ML so users can test out the full workflow.  @raimis do you think the other two packages are also in a suitable state for beta testers to use them, or are there changes we should get in first?

@tmarkland has a strong preference for calling this release 8.0.  Is that ok with everyone?

**Comment by raimis** [comment]
> It would be nice to have coordinated versions of OpenMM, OpenMM-Torch, and OpenMM-ML so users can test out the full workflow. @raimis do you think the other two packages are also in a suitable state for beta testers to use them, or are there changes we should get in first?

Regarding OpenMM-Torch, it is in a working state after the latest release (https://github.com/openmm/openmm-torch/releases/tag/v0.7). I'm not sure what is the status of OpenMM-ML.

> @tmarkland has a strong preference for calling this release 8.0. Is that ok with everyone?
I'm fine.

**Comment by peastman** [comment]
@jchodera given the above comments are you ok with moving forward?  Everyone else has given their ok.  If I don't hear from you by the end of the week, I'll start building the beta.

I think OpenMM-ML is in reasonable shape.  I just made a PR filling in the README.  Once that's merged, I think we can create a 1.0beta release and then build conda packages for it.  It would be nice if we could finish up one of the two PRs that adds NNPOps integration first, but since that only affects speed, not features, I don't think it's essential.

**Comment by zhang-ivy** [comment]
@peastman : What's the status of the beta release? Has it been built / Is it available on conda-forge? If not, when do you expect it to be?

**Comment by dominicrufa** [comment]
@peastman, I can update the `openmm-ml` test and push the `nnpops` bits tomorrow.

**Comment by peastman** [comment]
Excellent, thanks.

**Comment by jchodera** [comment]
Would be great to get https://github.com/openmm/openmm/pull/3386 in before we cut the `beta` so that we can have people send us their benchmarks in a standard JSON format!

**Comment by peastman** [comment]
It looks like #3386 is going to take a while yet, so I don't want to hold up the beta for it.

**Comment by jchodera** [comment]
I'm done with that PR, so you can move at your own pace to finish it
however you like. I just think it would be very useful for us to have some
way to capture submitted benchmarks in a machine-readable way.

**Comment by peastman** [comment]
The packages are now online!  Next we need to build corresponding packages for the `openmm-torch` and `openmm-ml` packages (and possibly `nnpops` as well)?  They should be able to generally follow [the same procedure](https://github.com/openmm/openmm/blob/master/devtools/CONDA_FORGE_RELEASES.md) used for OpenMM beta and rc builds.  The critical things are that they should be pushed to the `openmm_rc` label instead of `main`, and they should list the OpenMM 8.0 beta as a dependency.  They won't work correctly with 7.7.

**Comment by peastman** [comment]
I created a PR at https://github.com/conda-forge/openmm-torch-feedstock/pull/23 to build the beta package for OpenMM-Torch.  But it looks like we've never set it up to build on any platform other than Linux x86 and Mac x86.  I'm not sure exactly what needs to be done to make it build packages for other platforms.  And then we need to set up a feedstock to build packages for OpenMM-ML.

@mikemhenry and @raimis can you help?  This is pushing the limits of my experience building conda packages.

**Comment by mikemhenry** [comment]
@peastman what platforms do we want to target? Parity with openmm or a subset?

**Comment by peastman** [comment]
As many as possible, but of course it's limited by what PyTorch supports.  The conda-forge package includes ARM Mac as well.  The official packages on the `pytorch` channel also include Windows, but I don't think there's any way we can build against that on conda-forge. :(

**Comment by egallicc** [comment]
A personal note about HIP if you are still considering it. We have been playing around with the StreamHPC implementation  https://github.com/StreamHPC/openmm and https://github.com/StreamHPC/openmm-hip. We were able to build and test it on an AMD card with minor effort. We were also able to port the ATM Meta Force plugin to HIP via the Common platform (great invention btw), again, with very minor effort (https://github.com/Gallicchio-Lab/openmm-atmmetaforce-plugin/tree/hip). 

We have not done detailed benchmarking, but the performance appears unexpectedly good, probably ~3X better than with OpenCL. We are getting more throughput on an inexpensive and power-moderate AMD RX6750XT card than on a CUDA/V100 accelerator on XSEDE's Expanse.

It would be great to have HIP support in a stable OpenMM release soon.

**Comment by peastman** [comment]
Two months later, I think we finally have the problems with building the PyTorch plugin on conda-forge resolved!  Which means we can move forward with the beta release.  I believe these are the next steps.

1. I'll do a new beta build of OpenMM, since there have been a lot of code changes since I built the first one.
2. We need suitable conda packages for NNPOps and OpenMM-ML that work correctly with the OpenMM beta (either existing packages or beta packages coordinated with it).  @mikemhenry and @raimis what needs to be done for this?
3. @mikemhenry suggested creating a hosted environment file to make it easy for users to install appropriate beta versions of all the packages.  Can you describe exactly how that would be done?  (Consider not just how it will work for this release, but also future releases.)

**Comment by peastman** [comment]
@mikemhenry @raimis what's the status of the things listed above?  The OpenMM and OpenMM-Torch packages have been built, so those are all that's blocking the beta.

**Comment by mikemhenry** [comment]
I'll make a PR that details how to create the hosted environment so it will be easier for people to try the beta. 

Last I checked `NNPOps` needed an OSX build. 

I can have a status report put together tomorrow morning that triages any blockers.

**Comment by peastman** [comment]
Great, thank you!

**Comment by mikemhenry** [comment]
Okay so for `NNPOps`:
 - [ ] We need to get OSX builds working 
       * https://github.com/openmm/NNPOps/pull/63
       * https://github.com/conda-forge/nnpops-feedstock/pull/11
I tried getting builds on conda-forge working, but IMHO we should get OSX CI working before we try and get the OSX build on conda-forge working. I am also not sure if getting `NNPOps` working on OSX is worthwhile, is `NNPOps` useful without GPUs? I will admit that I am not involved with the development of `NNPOps` so I am not sure if 1) not having OSX builds is a blocker for the beta and 2) if OSX builds are useful for end users.

For `openmm-ml`:

We do have the package on [conda-forge](https://anaconda.org/conda-forge/openmm-ml/files) (with the openmm_rc label) but I have not personally tested the package. I do think we should get CI working on the main repo.
 - [ ] Add unit tests https://github.com/openmm/openmm-ml/issues/9 (I took a stab at solving this with https://github.com/openmm/openmm-ml/pull/40)

For the hosted environment file:
 - [ ] I've made this PR https://github.com/openmm/openmm/pull/3780

Just to make sure I don't miss any packages, is this the full list?
  * `openmm`
  * `openmm-ml`
  * `openmm-torch`
  * `NNPOps`

**Comment by peastman** [comment]
I think everything in OpenMM-ML is written to use NNPOps if it's installed, but not require it?  If so, no Mac builds should be ok.  It may even be faster without it, since PyTorch on Mac does use the GPU.

> Just to make sure I don't miss any packages, is this the full list?

Looks good.

**Comment by peastman** [comment]
Here's our current status.  We have builds of OpenMM, NNPOps, OpenMM-Torch, and OpenMM-ML.  However trying to install them fails in many situations.

- Installing on Mac always fails.  That's because we don't have NNPOps builds, and OpenMM-ML lists it as a dependency, even though it isn't really needed.  @mikemhenry is creating a new OpenMM-ML build for Mac that doesn't require it.
- Installing on Linux works with PyTorch 1.12, but fails with 1.11.  I don't understand why.  Neither NNPOps nor OpenMM-Torch specifies a version requirement for it, but somehow that particular version seems to be required.
- We have no Windows builds of NNPOps, OpenMM-Torch, or OpenMM-ML, because conda-forge doesn't provide PyTorch builds for Windows and we aren't allowed to build against packages from any other source.

**Comment by mikemhenry** [comment]
> * Installing on Mac always fails.  That's because we don't have NNPOps builds, and OpenMM-ML lists it as a dependency, even though it isn't really needed.  @mikemhenry is creating a new OpenMM-ML build for Mac that doesn't require it.

This is done now with https://github.com/conda-forge/openmm-ml-feedstock/pull/1 (will need to test once the build lands on the conda-forge CDN)



> * Installing on Linux works with PyTorch 1.12, but fails with 1.11.  I don't understand why.  Neither NNPOps nor OpenMM-Torch specifies a version requirement for it, but somehow that particular version seems to be required.

Is it important to support 1.11? Do we want to devote some time to fix this? Else I can make new release that specify 1.12 (or higher).

> We have no Windows builds of NNPOps, OpenMM-Torch, or OpenMM-ML, because conda-forge doesn't provide PyTorch builds for Windows and we aren't allowed to build against packages from any other source.

I am working on this here: https://github.com/conda-forge/pytorch-cpu-feedstock/pull/134 It looks like I might have to build the packages manually and upload them to conda-forge...

**Comment by peastman** [comment]
> Is it important to support 1.11?

It would certainly be useful.  I'm currently trying to use a model created with TorchMD-Net, and it doesn't yet support 1.12.  There likely are other packages out there that also don't support it yet.

> It looks like I might have to build the packages manually and upload them to conda-forge...

They let you do that???  I could have saved a month of work on the Mac builds if I could have just built them on my own computer rather than dealing with all the quirks in their cross compiling system!

**Comment by mikemhenry** [comment]
> It would certainly be useful. I'm currently trying to use a model created with TorchMD-Net, and it doesn't yet support 1.12. There likely are other packages out there that also don't support it yet.

I will look into then!

> They let you do that??? I could have saved a month of work on the Mac builds if I could have just built them on my own computer rather than dealing with all the quirks in their cross compiling system!

Only as a last resort https://github.com/conda-forge/cfep/blob/main/cfep-03.md I've only done it when it would take too many hours for CI to build the package.

**Comment by peastman** [comment]
@mikemhenry can I get your help with the OpenMM-ML builds?  I'm completely stuck.  We currently have no builds either for ARM Mac or Python 3.10 on any platform, and my attempt at switching back to noarch in https://github.com/conda-forge/openmm-ml-feedstock/pull/3 isn't working.

**Comment by mikemhenry** [comment]
Yes, I think for the python 3.10 builds we will need to make sure we have one for openmm that targets the rc label we are using (we might be doing that but it's the first thing that came to mind)

**Comment by jchodera** [comment]
Should we also update the [major and minor build numbers](https://github.com/openmm/openmm/blob/a39fa14acbb8eb00fb2dcb8f4e6ba0bc9b2564a9/CMakeLists.txt#L166-L168) so that OpenMM doesn't still report 7.7.0?

**Comment by mikemhenry** [comment]
> Should we also update the [major and minor build numbers](https://github.com/openmm/openmm/blob/a39fa14acbb8eb00fb2dcb8f4e6ba0bc9b2564a9/CMakeLists.txt#L166-L168) so that OpenMM doesn't still report 7.7.0?

That is a good idea, I'm guessing Peter is waiting to do that last, but now that we are getting to close to releasing these new builds, the old numbers will confuse people.

**Comment by jchodera** [comment]
The 8.0 beta reports the wrong version number right now:
```
>>> import openmm
>>> print(openmm.version.version)
7.7.0.dev-130124a
```

**Comment by jchodera** [comment]
Installing the 8.0 beta clearly picks up the beta:
```
$ conda activate openmm-8.0-beta
...
  openmm            8.0.0beta  py310hf68b541_1      conda-forge/label/openmm_rc/linux-64      12 MB
...
```
but then then shows the wrong version:
```
$ conda activate openmm-8.0-beta
(openmm-8.0-beta) [chodera@lilac:chodera]$ conda list openmm
# packages in environment at /lila/home/chodera/miniconda/envs/openmm-8.0-beta:
#
# Name                    Version                   Build  Channel
openmm                    7.7.0                    pypi_0    pypi
```
and in Python reports the wrong version number (but the right commit):
```
$ python
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import openmm
>>> print(openmm.version.version)
7.7.0.dev-a056d5a
```
Maybe we should create a checklist to make sure we update these things when we cut a beta?
```

**Comment by atom-moyer** [comment]
As related to this issue, what is the current recommended way to install the openmm-ml stack? What versions of `openmm` `openmm-ml` `openmm-torch` `NNPOps` should we try to install?

**Comment by peastman** [comment]
We're trying to get beta packages built for all of them, but we're having a lot of problems.  We've pretty much been stuck on fighting with conda-forge issues for a few months now. :(  At the moment, installing the latest revision of each one from source is probably the most reliable way.

**Comment by jchodera** [comment]
This is another issue with not using semantic versioning: We are not
clearly communicating with the developer community that, as a whole, as
adopted semantic versioning as a clear way of communicating when API
compatibility is expected.

**Comment by giadefa** [comment]
7.8. I would like to have a paper that comes out at the same time of 8.0.

**Comment by mikemhenry** [comment]
This should work at least for linux:
`mamba env create mmh/openmm-8-beta-linux`
Which will make an evn with 
`'openmm==8.0.0beta', 'openmm-ml==1.0beta', 'openmm-torch==1.0beta', 'nnpops==0.2'`

@peastman Can you point me to a script that would test this whole stack?

**Comment by peastman** [comment]
Testing on Linux, it appears to install the correct versions of everything:

```
$ mamba env create mmh/openmm-8-beta-linux
conda-forge/label/openmm [====================] (00m:00s) Done
conda-forge/label/openmm [====================] (00m:00s) Done
pkgs/r/linux-64          [====================] (00m:00s) No change
conda-forge/label/openmm [====================] (00m:00s) Done
pkgs/r/noarch            [====================] (00m:00s) No change
pkgs/main/noarch         [====================] (00m:00s) Done
conda-forge/label/openmm [====================] (00m:00s) Done
conda-forge/noarch       [====================] (00m:08s) Done
pkgs/main/linux-64       [====================] (00m:03s) Done
conda-forge/linux-64     [====================] (00m:17s) Done


Looking for: ['openmm==8.0.0beta', 'openmm-ml==1.0beta', 'openmm-torch==1.0beta', 'nnpops==0.2']


Transaction

  Prefix: /home/peastman/miniconda3/envs/openmm-8-beta-linux

  Updating specs:

   - openmm==8.0.0beta
   - openmm-ml==1.0beta
   - openmm-torch==1.0beta
   - nnpops==0.2


WARNING Missing file from package cache: "/home/peastman/miniconda3/pkgs/_openmp_mutex-4.5-2_gnu/lib/libgomp.so.1"
WARNING Missing file from package cache: "/home/peastman/miniconda3/pkgs/libblas-3.9.0-16_linux64_mkl/lib/libblas.so"
WARNING Missing file from package cache: "/home/peastman/miniconda3/pkgs/libcblas-3.9.0-16_linux64_mkl/lib/libcblas.so"
WARNING Missing file from package cache: "/home/peastman/miniconda3/pkgs/liblapack-3.9.0-16_linux64_mkl/lib/liblapack.so"
  Package                  Version  Build                     Channel                                          Size
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  _libgcc_mutex                0.1  conda_forge               conda-forge/linux-64                           Cached
  _openmp_mutex                4.5  2_gnu                     conda-forge/linux-64                           Cached
  brotlipy                   0.7.0  py310h5764c6d_1004        conda-forge/linux-64                           342 KB
  bzip2                      1.0.8  h7f98852_4                conda-forge/linux-64                           Cached
  c-ares                    1.18.1  h7f98852_0                conda-forge/linux-64                           Cached
  ca-certificates        2022.9.24  ha878542_0                conda-forge/linux-64                           Cached
  cached-property            1.5.2  hd8ed1ab_1                conda-forge/noarch                             Cached
  cached_property            1.5.2  pyha770c72_1              conda-forge/noarch                             Cached
  certifi                2022.9.24  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  cffi                      1.15.1  py310h255011f_0           conda-forge/linux-64                           231 KB
  charset-normalizer         2.1.1  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  cryptography              37.0.1  py310h9ce1e76_0           pkgs/main/linux-64                               2 MB
  cudatoolkit               11.7.0  hd8887f6_10               conda-forge/linux-64                           Cached
  cudnn                   8.4.1.50  hed8a83a_0                conda-forge/linux-64                           Cached
  h5py                       3.7.0  nompi_py310h416281c_101   conda-forge/linux-64                             1 MB
  hdf5                      1.12.2  nompi_h4df4325_100        conda-forge/linux-64                           Cached
  idna                         3.4  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  importlib-metadata        4.11.4  py310hff52083_0           conda-forge/linux-64                            33 KB
  importlib_metadata        4.11.4  hd8ed1ab_0                conda-forge/noarch                             Cached
  intel-openmp            2022.1.0  h9e868ea_3769             pkgs/main/linux-64                             Cached
  keyutils                   1.6.1  h166bdaf_0                conda-forge/linux-64                           Cached
  krb5                      1.19.3  h08a2579_0                conda-forge/linux-64                           Cached
  lark-parser               0.12.0  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  ld_impl_linux-64            2.38  h1181459_1                pkgs/main/linux-64                             Cached
  libblas                    3.9.0  16_linux64_mkl            conda-forge/linux-64                           Cached
  libcblas                   3.9.0  16_linux64_mkl            conda-forge/linux-64                           Cached
  libcurl                   7.85.0  h2283fc2_0                conda-forge/linux-64                           351 KB
  libedit             3.1.20210910  h7f8727e_0                pkgs/main/linux-64                             Cached
  libev                       4.33  h516909a_1                conda-forge/linux-64                           Cached
  libffi                     3.4.2  h7f98852_5                conda-forge/linux-64                           Cached
  libgcc-ng                 12.1.0  h8d9b700_16               conda-forge/linux-64                           Cached
  libgfortran-ng            12.1.0  h69a702a_16               conda-forge/linux-64                           Cached
  libgfortran5              12.1.0  hdcd56e2_16               conda-forge/linux-64                           Cached
  libgomp                   12.1.0  h8d9b700_16               conda-forge/linux-64                           Cached
  liblapack                  3.9.0  16_linux64_mkl            conda-forge/linux-64                           Cached
  libnghttp2                1.47.0  hff17c54_1                conda-forge/linux-64                           Cached
  libnsl                     2.0.0  h7f98852_0                conda-forge/linux-64                           Cached
  libprotobuf               3.20.1  h6239696_4                conda-forge/linux-64                           Cached
  libsqlite                 3.39.4  h753d276_0                conda-forge/linux-64                           803 KB
  libssh2                   1.10.0  hf14f497_3                conda-forge/linux-64                           Cached
  libstdcxx-ng              12.1.0  ha89aaad_16               conda-forge/linux-64                           Cached
  libuuid                   2.32.1  h14c3975_1000             conda-forge/linux-64                           Cached
  libzlib                   1.2.12  h166bdaf_4                conda-forge/linux-64                            65 KB
  magma                      2.5.4  h6103c52_2                conda-forge/linux-64                           Cached
  mkl                     2022.1.0  hc2b9512_224              pkgs/main/linux-64                             Cached
  nccl                    2.14.3.1  h0800d71_0                conda-forge/linux-64                           Cached
  ncurses                      6.3  h27087fc_1                conda-forge/linux-64                           Cached
  ninja                     1.11.0  h924138e_0                conda-forge/linux-64                           Cached
  nnpops                       0.2  cuda112py310h85a0d14_4    conda-forge/linux-64                           998 KB
  numpy                     1.23.3  py310h53a5b5f_0           conda-forge/linux-64                             7 MB
  ocl-icd                    2.3.1  h7f98852_0                conda-forge/linux-64                           Cached
  ocl-icd-system             1.0.0  1                         conda-forge/linux-64                           Cached
  openmm                 8.0.0beta  py310h2996cf7_2           conda-forge/label/openmm_rc/linux-64            12 MB
  openmm-ml                1.0beta  pyh79ba5db_0              conda-forge/label/openmm_rc/noarch              14 KB
  openmm-torch             1.0beta  cuda112py310hcc28b43_1    conda-forge/label/openmm-torch_rc/linux-64     153 KB
  openssl                    3.0.5  h166bdaf_2                conda-forge/linux-64                           Cached
  packaging                   21.3  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  pip                       22.2.2  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  pycparser                   2.21  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  pyopenssl                 22.0.0  pyhd8ed1ab_1              conda-forge/noarch                             Cached
  pyparsing                  3.0.9  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  pysocks                    1.7.1  py310hff52083_5           conda-forge/linux-64                            28 KB
  python                    3.10.6  ha86cf86_0_cpython        conda-forge/linux-64                            29 MB
  python_abi                  3.10  2_cp310                   conda-forge/linux-64                             4 KB
  pytorch                   1.11.0  cuda112py310h51fe464_202  conda-forge/linux-64                           490 MB
  readline                   8.1.2  h0f457ee_0                conda-forge/linux-64                           Cached
  requests                  2.28.1  pyhd8ed1ab_1              conda-forge/noarch                             Cached
  setuptools                59.5.0  py310hff52083_0           conda-forge/linux-64                             1 MB
  setuptools-scm             6.3.2  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  setuptools_scm             6.3.2  hd8ed1ab_0                conda-forge/noarch                             Cached
  sleef                      3.5.1  h28343ad_2                conda-forge/linux-64                           Cached
  tk                        8.6.12  h27826a3_0                conda-forge/linux-64                           Cached
  tomli                      2.0.1  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  torchani                   2.2.2  cuda112py310h73d5bcf_5    conda-forge/linux-64                            22 MB
  typing_extensions          4.4.0  pyha770c72_0              conda-forge/noarch                              29 KB
  tzdata                     2022e  h191b570_0                conda-forge/noarch                             118 KB
  urllib3                  1.26.11  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  wheel                     0.37.1  pyhd8ed1ab_0              conda-forge/noarch                             Cached
  xz                         5.2.6  h166bdaf_0                conda-forge/linux-64                           Cached
  zipp                       3.9.0  pyhd8ed1ab_0              conda-forge/noarch                              13 KB
  zlib                      1.2.12  h166bdaf_4                conda-forge/linux-64                            92 KB

  Summary:

  Install: 81 packages

  Total download: 566 MB
```

To test it, I ran the TestMLPotential.py script from https://github.com/openmm/openmm-ml/tree/main/test.

```
$ python TestMLPotential.py 
/home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/torchani/__init__.py:55: UserWarning: Dependency not satisfied, torchani.ase will not be available
  warnings.warn("Dependency not satisfied, torchani.ase will not be available")
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
/home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/torchani/resources/
Downloading ANI model parameters ...
/home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/torchani/resources/
E
======================================================================
ERROR: testCreateMixedSystem (__main__.TestMLPotential)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/peastman/workspace/openmm-ml/test/TestMLPotential.py", line 18, in testCreateMixedSystem
    mmContext = mm.Context(mmSystem, mm.VerletIntegrator(0.001), platform)
  File "/home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/openmm/openmm.py", line 3530, in __init__
    _openmm.Context_swiginit(self, _openmm.new_Context(*args))
openmm.OpenMMException: Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)

----------------------------------------------------------------------
Ran 1 test in 46.881s
```

I believe that error is caused by conda picking the wrong CUDA toolkit.  It installed 11.7, even though my driver only supports 11.6.  So I downgraded (`mamba install -c conda-forge cudatoolkit=11.6`) and tried again.

```
$ python TestMLPotential.py 
Traceback (most recent call last):
  File "TestMLPotential.py", line 1, in <module>
    import openmm as mm
ModuleNotFoundError: No module named 'openmm'
```

At this point I don't understand what's going on and why it can't load the module.  I'll keep investigating.

**Comment by peastman** [comment]
For some reason, installing the CUDA toolkit caused conda to get confused.  Even though I had the environment activated, it was running the python from the base environment instead.  Deactivating and reactivating it fixed that.  Strange!

After doing that I had to install pytest, since it wasn't included in the environment, and then found a bug in the test program that kept it from running.  (I'll send a PR with the fix in just a moment.)  But once I did all that, I was able to successfully run it!

```
$ pytest TestMLPotential.py 
===================================================== test session starts ======================================================
platform linux -- Python 3.10.6, pytest-7.1.3, pluggy-1.0.0
rootdir: /home/peastman/workspace/openmm-ml
collected 8 items                                                                                                              

TestMLPotential.py ........                                                                                              [100%]

======================================================= warnings summary =======================================================
../../../miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/openmm/app/gromacstopfile.py:45
  /home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/openmm/app/gromacstopfile.py:45: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    import distutils.spawn

test/TestMLPotential.py::TestMLPotential::testCreateMixedSystem[nnpops-0]
  /home/peastman/miniconda3/envs/openmm-8-beta-linux/lib/python3.10/site-packages/torchani/__init__.py:55: UserWarning: Dependency not satisfied, torchani.ase will not be available
    warnings.warn("Dependency not satisfied, torchani.ase will not be available")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================ 8 passed, 2 warnings in 56.12s ================================================
```

**Comment by peastman** [comment]
Installing on Mac, the test fails because torchani isn't getting installed.

```
================================================================ short test summary info ================================================================
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[nnpops-0] - ModuleNotFoundError: No module named 'torchani'
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[nnpops-1] - ModuleNotFoundError: No module named 'torchani'
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[nnpops-2] - ModuleNotFoundError: No module named 'torchani'
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[torchani-0] - ModuleNotFoundError: No module named 'torchani'
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[torchani-1] - ModuleNotFoundError: No module named 'torchani'
FAILED TestMLPotential.py::TestMLPotential::testCreateMixedSystem[torchani-2] - ModuleNotFoundError: No module named 'torchani'
=================================================================== 6 failed in 1.50s ===================================================================
```

I think the recipe for `openmm-ml` should list it as a dependency?

**Comment by mikemhenry** [comment]
I will add that missing `torchani` dep https://github.com/conda-forge/openmm-ml-feedstock/pull/4. Does if work if you install `torchani` manually into the environment?

**Comment by peastman** [comment]
Yes.  After manually installing `torchani` the test runs successfully.

**Comment by mikemhenry** [comment]
@peastman that's great! Once the build lands on the CDN, then we should be good? I'll make a PR to update the hosted environments with the versions pinned.

**Comment by jchodera** [comment]
Hooray!

How do we want people to test this once they've installed it? If we want them to run the pytest, should we also add that to the recipe?

**Comment by jchodera** [comment]
Another idea: Do we want folks to run an ML/MM benchmark that reports ns/day for a protein:ligand system to compare to MM? If so, where should this live?

**Comment by mikemhenry** [comment]
I can add pytest if that is the best way for people to test it, it might be worth having a script we can call like `python -m openmm.test_ml_stack` or something that works like `python -m openmm.testInstallation`

**Comment by jchodera** [comment]
Ideally, we would have something like `python -m openmm.benchmarks.ml_benchmarks` that would also report useful benchmark data as well. But for the short term, just allowing people to run the pytests for the beta would be extremely helpful.

**Comment by peastman** [comment]
Running the test cases will help people verify that everything is installed correctly, but it may not be very educational for them.  It would be great if we could also put together a very brief script demonstrating how to use it.  At present we don't really have any documentation for it beyond what's in the README at https://github.com/openmm/openmm-ml.

**Comment by mikemhenry** [comment]
Okay new PR with versions pinned + added pytest (let me know if there is anything else you think would be worth including, people can always install what they need into the envs)
https://github.com/openmm/openmm/pull/3808

**Comment by peastman** [comment]
Thanks!  The PR is merged.  If you can post the environments to anaconda, I'll give them a final test and then we can announce the beta!

**Comment by peastman** [comment]
How do we add you?  I'm not familiar with how teams work.

**Comment by peastman** [comment]
It says I don't have permission to view that page.  Apparently I'm not a member either.  It also looks like that organization was only created a couple of years ago, and has never been used for anything.  Does anyone know who created it and who are the members?

**Comment by Lnaden** [comment]
Where exactly are you trying to upload these environments too? The `https://anaconda.org/openmm` group? I think that group was made as a placeholder back in 2020, but I don't know who made it. We've never done anything with it at least. Do you have to upload the environment files to the `openmm` group?

**Comment by peastman** [comment]
It probably doesn't matter much which group it's in.  This is just a beta.  If you want to put them at mmh, I think that's fine for the moment.  If it turns out we'll be doing this a lot, we can figure out a better location later.

**Comment by mikemhenry** [comment]
Okay! Then we should be good to go:
Linux:
`mamba env create mmh/openmm-8-beta-linux`
Mac:
`mamba env create mmh/openmm-8-beta-mac`

**Comment by peastman** [comment]
Thanks!  I just tested it, and of course Nothing Ever Works.  TestMLPotential.py reports lots of failures that seem to be related to torchani.

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/openmmml/mlpotential.py:267: in createMixedSystem
    self._impl.addForces(topology, newSystem, atomList, forceGroup, **args)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/openmmml/models/anipotential.py:135: in addForces
    module = torch.jit.script(aniForce)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_script.py:1286: in script
    return torch.jit._recursive.create_script_module(
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_recursive.py:458: in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_recursive.py:520: in create_script_module_impl
    script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_script.py:615: in _construct
    init_fn(script_module)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_recursive.py:498: in init_fn
    scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_recursive.py:520: in create_script_module_impl
    script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/jit/_script.py:615: in _construct
    init_fn(script_module)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

script_module = RecursiveScriptModule(original_name=BuiltinEnsemble)

    def init_fn(script_module):
        # Initialize the ScriptModule:
        # 1. Copy the attributes/parameters/buffers from the original `nn_module` to the new ScriptModule.
        for name, (attr_type, is_param) in concrete_type.get_attributes().items():
            orig_value = getattr(nn_module, name)
            orig_value = orig_value.value if isinstance(orig_value, torch.jit.Attribute) else orig_value
>           cpp_module.setattr(name, orig_value)
E           RuntimeError: Could not cast attribute '_species_to_tensor' to type __torch__.torchani.utils.ChemicalSymbolsToInts (of Python compilation unit at: 0x6000009edb98): Could not cast attribute 'rev_species' to type Dict[Tensor, int]: Unable to cast H to Tensor
```

Has something changed since the last version?

**Comment by peastman** [comment]
It seems to be related to the python version.  If I force it to use 3.9 the tests pass.  If I let it install 3.10 they don't.

**Comment by peastman** [comment]
Actually it's the torchani version.  For some reason it installs different versions for the different Python versions.  With 2.2.2 it works.  WIth 2.2.0 it doesn't.  Can you add `torchani==2.2.2` to the environment files?

**Comment by mikemhenry** [comment]
I will fix the conda package as well, but for now I can quickly update the env files

**Comment by peastman** [comment]
Thanks!  It now works on Linux.  On Mac it usually segfaults, but occasionally I instead get an exception:

```
>       state = _openmm.Context_getState(self, types, enforcePeriodicBox, groups_mask)
E       openmm.OpenMMException: Could not allocate memory for Tensor SizesAndStrides!
E       Exception raised from allocateOutOfLineStorage at /Users/runner/miniforge3/conda-bld/pytorch-recipe_1664817728005/work/c10/core/impl/SizesAndStrides.h:279 (most recent call first):
E       frame #0: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 188 (0x11bdd96c0 in libc10.dylib)
E       frame #1: c10::TensorImpl::copy_generic_tensor_metadata(c10::TensorImpl const*, c10::TensorImpl*) + 452 (0x11bdc6fc4 in libc10.dylib)
...
```

I have no idea if this is something specific to my computer or a problem with the build.  Can someone else with a Mac please try out TestMLPotential and see if it works for you?

(By the way, please post a new comment rather than editing an existing one if you want something to be seen.  It doesn't send out emails when you edit an existing comment.  I've been waiting all this time for you to make the update, then happened to reload the page and discovered you had done it three hours ago!)

**Comment by peastman** [comment]
Forcing it to downgrade PyTorch to 1.11 makes the errors go away.  Upgrading back to 1.12.1 makes it start failing with a completely different error:

```
E       ImportError: dlopen(/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so, 0x0002): Library not loaded: '@rpath/libOpenMM.7.7.dylib'
E         Referenced from: '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so'
E         Reason: tried: '/Users/runner/work/1/s/recipe/libtorch/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/runner/work/1/s/recipe/libtorch/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/../../libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/runner/work/1/s/recipe/libtorch/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/runner/work/1/s/recipe/libtorch/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/../../libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/torch/lib/libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/bin/../lib/libOpenMM.7.7.dylib' (no such file), '/Users/peastman/miniconda3/envs/openmm-8-beta-mac/bin/../lib/libOpenMM.7.7.dylib' (no such file), '/usr/local/lib/libOpenMM.7.7.dylib' (no such file), '/usr/lib/libOpenMM.7.7.dylib' (no such file)

../../../miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/openmmtorch.py:15: ImportError
```

Notice that it's now looking for OpenMM 7.7.  That's despite the fact that I still have the correct versions of all libraries installed:

```
openmm                    8.0.0beta       py310hd6c29c8_2_khronos    conda-forge/label/openmm_rc
openmm-ml                 1.0beta            pyh79ba5db_1    conda-forge/label/openmm_rc
openmm-torch              1.0beta         cpu_py310h141eede_0    conda-forge/label/openmm-torch_rc
```

In fact, I have *never* installed any other version of them in this environment.  Yet somehow it's still referencing a different OpenMM version it shouldn't even know exists???

**Comment by peastman** [comment]
Immediately after creating the environment, the OpenMM-Torch library correctly links to OpenMM 8.0:

```
$ otool -L /Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so
/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so:
	@rpath/libOpenMM.8.0.dylib (compatibility version 8.0.0, current version 0.0.0)
	@rpath/libOpenMMTorch.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
```

I downgrade to PyTorch 1.11:

```
$ mamba install -c conda-forge/label/openmm_rc -c conda-forge/label/openmm-torch_rc -c conda-forge pytorch=1.11
```

It doesn't say it's going to change the OpenMM-Torch package, and in fact the library is still correct afterward:

```
$ otool -L /Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so
/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so:
	@rpath/libOpenMM.8.0.dylib (compatibility version 8.0.0, current version 0.0.0)
	@rpath/libOpenMMTorch.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
```

Now tell it to switch back to 1.12.1:

```
$ mamba install -c conda-forge/label/openmm_rc -c conda-forge/label/openmm-torch_rc -c conda-forge pytorch=1.12.1
```

This time it *does* say it's going to change the package, though just to a different build of the same version:

```
  Package         Version  Build                Channel                                           Size
────────────────────────────────────────────────────────────────────────────────────────────────────────
  Change:
────────────────────────────────────────────────────────────────────────────────────────────────────────

  - openmm-torch  1.0beta  cpu_py310h2455b05_1  conda-forge/label/openmm-torch_rc                     
  + openmm-torch  1.0beta  cpu_py310h141eede_0  conda-forge/label/openmm-torch_rc/osx-arm64     Cached
  - torchani        2.2.2  cpu_py310h1be3dbc_5  conda-forge                                           
  + torchani        2.2.2  cpu_py310hfaa7d71_6  conda-forge/osx-arm64                           Cached
```

And it now references the wrong OpenMM version:

```
$ otool -L /Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so
/Users/peastman/miniconda3/envs/openmm-8-beta-mac/lib/python3.10/site-packages/_openmmtorch.cpython-310-darwin.so:
	@rpath/libOpenMM.7.7.dylib (compatibility version 7.7.0, current version 0.0.0)
	@rpath/libOpenMMTorch.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
```

I don't know if this is connected to the segfault (the Mac build for PyTorch 1.12.1 is just generally messed up), or if these are two unrelated problems.

**Comment by peastman** [comment]
The build log at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=583323&view=results shows that after the most recent openmm-torch update (https://github.com/conda-forge/openmm-torch-feedstock/pull/27), several of the Mac builds failed.  Possibly that's related?  I'm going to rerun the failed builds and see if I can get them to work.

**Comment by peastman** [comment]
I told it to rerun the failed builds, and instead it's rerunning all builds.  Sigh.

**Comment by peastman** [comment]
After rerunning the full set of builds a couple of times, I hopefully have gotten everything to succeed.  I can see at https://anaconda.org/conda-forge/openmm-torch/files that five new Mac packages have gotten uploaded.  And both of the above problems appear to be fixed.  After installing the environment file, TestMLPotential passes, and it continues to pass if I switch to PyTorch 1.11 and then back to 1.12.1.

I think this means we may finally be ready to make the beta official!  Shall I post the release announcement?

**Comment by mikemhenry** [comment]
@peastman sorry about that edit! I've made this PR:
https://github.com/conda-forge/openmm-ml-feedstock/pull/5

I think we are good to make the beta official! We don't need to wait for that build/PR since we have that pin the environment file, but I thought it would be good to get into the recipe.

**Comment by jchodera** [comment]
Hooray!

@peastman What do you think about (1) cutting a GitHub release (as a preprelease) with release notes (features, bugfixes, behavior changes) for the beta, and (2) drafting the release announcement for feedback before announcing? We can announce tomorrow then?

**Comment by peastman** [comment]
I'll email out a draft of the release announcement.  Trying to discuss it here will add too much clutter to the discussion.

**Comment by peastman** [comment]
The [release announcement](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=15133&p=0&start=0&view=) is now up!  I'll create a github release to go with it momentarily.  We should expect the release candidate to follow in about a month.  There's a lot to do in that time.

If we can create a complete example showing how to use the ML features, that would be really useful.  It doesn't have to be anything complicated.  In fact, it's better if it isn't.  Just something to help people get started testing it out.

There's lots of updates to make to the documentation.  I'll get started on that.  I'll also do a sweep through the tracker to find bugs that should be fixed.

Building conda packages for the beta was traumatic.  In the process, I think some issues were swept under the carpet in the interest of getting something out.  Now is the time to figure out any remaining issues so that when we're ready for the release candidate, the process can be as smooth as possible.

Congratulations everyone on getting to this point!

**Comment by jchodera** [comment]
> Building conda packages for the beta was traumatic. In the process, I think some issues were swept under the carpet in the interest of getting something out. Now is the time to figure out any remaining issues so that when we're ready for the release candidate, the process can be as smooth as possible.

Would it make sense to set up a meeting with some of the conda-forge folks to review some of these issues? Are you on their [gitter](https://gitter.im/conda-forge/conda-forge.github.io), @peastman? (See [conda-forge contact info here](https://conda-forge.org/docs/orga/getting-in-touch.html).)

**Comment by peastman** [comment]
It's already been close to two weeks since we announced the beta, which means the release candidate is coming up fast.  I marked issues with the [8.0 milestone](https://github.com/openmm/openmm/milestone/5) that I think need to be fixed in it.  Take a look and see if there's anything else you think should be included.

Most of them should be straightforward.  The two exceptions are #3815 (problems with the CHARMM Drude force field) and #3812 (errors using NNPOps with PyTorch 1.12).  Both of those might take a lot of debugging.  We should focus on fixing them as soon as possible, since they're the biggest risk to the schedule.  I'm working on #3815 now.

**Comment by jchodera** [comment]
> It's already been close to two weeks since we announced the beta, which means the release candidate is coming up fast.

I think we want to be cautious here to avoid repeating the issues with OpenMM 7.6. Have we heard confirmation from many developers that use OpenMM saying the beta is working for them? If not, we should use this time to do more outreach and see if folks are testing with the beta.

**Comment by jchodera** [comment]
I've [tweeted a request to test against OpenMM 8 beta](https://twitter.com/openmm_toolkit/status/1587519998251003905) and also pinged the OpenFF devs. 

@peastman : Please reach out to more dev communities if you can.

**Comment by peastman** [comment]
We won't release it until we're comfortable it's ready.  But in the mean time, we need to be moving quickly to fix any problems we know about.

**Comment by jchodera** [comment]
We're just working our way through some issues now:
* https://github.com/choderalab/openmmtools/pull/628

**Comment by jchodera** [comment]
@peastman : We're tracking another issue with the OpenMM 8 beta here:
https://github.com/openmm/openmm/issues/3832

I've updated the previous post too.

**Comment by peastman** [comment]
We're getting much closer.  At this point, I believe the only significant unresolved issue is https://github.com/openmm/openmm-torch/issues/88.  We build openmm-torch against both pytorch 1.11 and 1.12.  When installing with conda, it installs the wrong one.  You get one version of pytorch, and the openmm-torch package that was built against the other one.  This leads to a segfault.  It does not happen when installing with mamba.  In that case, you get the right package.

I don't know whether this is a bug in conda, or whether there's a problem in how we specify the constraints in the recipe (which I don't entirely understand).  If anyone is able to help with figuring this out, it would be really appreciated.  If all else fails, we could just decide we'll only support pytorch 1.12, but it would be better if we can give people flexibility.

**Comment by peastman** [comment]
Also, pytorch 1.13 is now out.

**Comment by bdenhollander** [comment]
@peastman I attempted to test the beta on Windows but all the Python builds request a version of `numpy` that requires Python 3.10. I see some discussion about [3.10](https://github.com/openmm/openmm/issues/3456) and there is still no win-64 py310 listed on [conda forge](https://anaconda.org/conda-forge/openmm/files?version=8.0.0beta). Are the issues blocking Python 3.10 on Windows resolved? Is it possible to relax the minimum `numpy` version or Python 3.7-3.9?
```
conda create -c "conda-forge/label/openmm_rc" --name openmm-8.0 python=3.7 openmm=8.0.0beta

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package python conflicts for:
openmm=8.0.0beta -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.7
openmm=8.0.0beta -> numpy[version='>=1.20.3,<2.0a0'] -> python[version='>=3.10,<3.11.0a0']
```
```
conda create -c "conda-forge/label/openmm_rc" --name openmm-8.0 python=3.8 openmm=8.0.0beta

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package python conflicts for:
openmm=8.0.0beta -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.8
openmm=8.0.0beta -> numpy[version='>=1.20.3,<2.0a0'] -> python[version='>=3.10,<3.11.0a0']
```
```
conda create -c "conda-forge/label/openmm_rc" --name openmm-8.0 python=3.9 openmm=8.0.0beta

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package python conflicts for:
openmm=8.0.0beta -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
openmm=8.0.0beta -> numpy[version='>=1.20.3,<2.0a0'] -> python[version='>=3.10,<3.11.0a0']
python=3.9
```

**Comment by peastman** [comment]
Try adding the flag `-c conda-forge` to the command line.  I think it may just be missing that channel.

**Comment by bdenhollander** [comment]
> Try adding the flag `-c conda-forge` to the command line. I think it may just be missing that channel.

That doesn't help. I also tried `-c "conda-forge/label/openmm_dev"` and it has the same Python incompatibility issues as `-c "conda-forge/label/openmm_rc"`.

```
conda create -c conda-forge --name openmm-8.0 python=3.10 openmm=8.0.0beta
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - openmm=8.0.0beta

Current channels:

  - https://conda.anaconda.org/conda-forge/win-64
  - https://conda.anaconda.org/conda-forge/noarch
  -  https://repo.anaconda.com/pkgs/main/win-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/win-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/msys2/win-64
  - https://repo.anaconda.com/pkgs/msys2/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

**Comment by peastman** [comment]
Neither of those is the correct command line to use.  Try installing it with exactly this command:

```
conda install -c conda-forge/label/openmm_rc -c conda-forge openmm
```

**Comment by bdenhollander** [comment]
> Neither of those is the correct command line to use. Try installing it with exactly this command:
> 
> ```
> conda install -c conda-forge/label/openmm_rc -c conda-forge openmm
> ```

Thank you! That found a working combination of packages running on Python 3.8.

**Comment by peastman** [comment]
Happy new year everyone!  I'd like to plan on building the release candidate next week.  That means the few remaining issues need to get resolved this week.

@jchodera can you check the changes in #3386?

@mikemhenry can you finish up https://github.com/conda-forge/openmm-torch-feedstock/pull/31 so we can test it and find out whether it fixes the pytorch problems?

We're getting very close.

**Comment by peastman** [comment]
I'm building the release candidate at https://github.com/conda-forge/openmm-feedstock/pull/95.  Once it is done, we'll need to create release candidate builds of openmm-torch and openmm-ml that depend on it.

@raimis do we need a new build of NNPOps?

**Comment by raimis** [comment]
> @raimis do we need a new build of NNPOps?

I think the don't need to rebuild.

**Comment by peastman** [comment]
Updated builds are done for NNPOps, OpenMM, OpenMM-Torch, and OpenMM-ML.  One last step before we announce it: @mikemhenry can you update the environment file so it installs the new builds?

**Comment by mikemhenry** [comment]
@peastman Will do!

**Comment by mikemhenry** [comment]
Envs are posted here:
```
environment located at:
https://anaconda.org/mmh/openmm-8-rc1-linux

environment located at:
https://anaconda.org/mmh/openmm-8-rc1-mac
```

I also pushed them to PR #3808

**Comment by peastman** [comment]
The environment has the wrong version of NNPOps.  Can you update it to install 0.3 instead of 0.2?

**Comment by peastman** [comment]
New packages are built with the fix from #3912.  @mikemhenry we're just waiting for you to fix the problem in the environment file.  Then we can announce it.

**Comment by peastman** [comment]
Thanks @mikemhenry!  I confirmed that `conda env create mmh/openmm-8-rc1-linux` installs the right versions of everything and that `TestMLPotential.py` runs correctly.  That means we're ready to call this an official release candidate.  I'll post an announcement on the forum.  Feel free to notify your followers on Twitter (does it still exist?) or wherever else you think will be good for reaching people.  We want to encourage people to run their actual workflows with it and make sure everything is ok.  The more testing we can get the better.

If no problems are found, we'll plan to make it an official release in 1-2 weeks.

**Comment by mikemhenry** [comment]
I updated the PR+pushed the envs to the cloud

**Comment by peastman** [comment]
I'm building an updated release candidate to get the fixes from #3923.  They're very minor, localized changes that shouldn't impact any of the downstream packages.  They just fix the Python wrappers for two fairly obscure methods.  They're mainly important because they impact ForceBalance.

**Comment by mikemhenry** [comment]
@peastman I will update the hosted envs, is the only new build for openmm? I'll need to make new builds of the downstream packages so they pull in `8.0.0rc2` instead of the old one.

**Comment by peastman** [comment]
I don't think that's necessary.  As noted above, the only change was unrelated to ML.  Lee-Ping already confirmed it fixed the problem.  For everything else, it's fine if people are testing rc1.

**Comment by peastman** [comment]
There have been no further reports of problems.  That means it's time to start building the release!  I'll create a PR for the OpenMM package, then we can move on to the other packages.  Hopefully we can release everything tomorrow.

**Comment by jchodera** [comment]
Can we hold off until at least Monday? We are still tracking down an issue with running perses / openmmtools with the latest rc.

**Comment by peastman** [comment]
What issue?

**Comment by jchodera** [comment]
I'll ask @ijpulidos and @mikemhenry to summarize their findings, but we're still trying to complete a successful set of local GPU tests of our tools. We have not yet done this.
The other outstanding issues I know of are
1. [This issue with Python subclasses of OpenMM classes](https://github.com/choderalab/openmmtools/pull/649#issuecomment-1404318813). I think this is an issue with a change in behavior affecting openmmtools but we haven't fully tracked this down yet.
2. The [difference in computed results](https://github.com/openmm/openmm/issues/3832#issuecomment-1404088712) we see, which we suspect to be an issue with the OpenEye toolkit import and are working with them to resolve.

There may be others I don't have visibility into right now, but in any case, waiting until Monday seems prudent.

@ijpulidos @mikemhenry : Can you make sure to summarize your findings?

**Comment by peastman** [comment]
It's Monday and you haven't posted anything further, so I'm assuming it's ok to move forward with the release.  If I don't hear from you in the next hour or so, I'll merge the PR to build the packages.

**Comment by ijpulidos** [comment]
As far as I can tell the findings are correctly summarized in the last comment from @jchodera . The behavior on how global parameters changed in commit https://github.com/openmm/openmm/commit/9d9ffb016b94fefe2f6aba5468173cf20ebb7433 and we have `CustomNonbondedForce` objects that define different default values for the same global parameter, which don't seem to be allowed now in Openmm8.

I still have to dig deeper what are the consequences of this behavior in terms of `openmmtools`. It is only a single specific test that is failing and maybe we just need to refactor how the `RestorableOpenMMObject` works.

**Comment by peastman** [comment]
That isn't really a change in behavior.  It was always wrong and produced undefined results.  It's just that it now tells you you're doing something wrong, where before you would silently get different results than you expected.

**Comment by jchodera** [comment]
@peastman I think we are safe to move forward.
Can you point me to the complete release notes to tweet once things have posted?

**Comment by peastman** [comment]
All packages are built.  If I install OpenMM and OpenMM-Torch with

```
mamba install -c conda-forge openmm-torch
```

it installs the correct packages and all tests pass.

If instead I tell it to install OpenMM-ML with

```
mamba install -c conda-forge openmm-ml
```

then it incorrectly installs a CPU build of PyTorch:

```
  pytorch                   1.12.1  cpu_py310h75c9ab6_0      conda-forge/linux-64      62 MB
```

This causes all tests involving PyTorch to fail with the error

> E   ImportError: libtorch_cuda.so: cannot open shared object file: No such file or directory

Any idea what causes that?

**Comment by sef43** [comment]
I don't get this error, if I do `mamba install -c conda-forge openmm-ml` in a new conda environment it correctly installs the latest versions of openmm, openmm-torch, openmm-ml, and pytorch 1.13.1 cuda

**Comment by peastman** [comment]
The behavior seems to be really erratic depending on the python version, whether you use conda or mamba, whether you install openmm-ml when you create the environment, and probably some other things.  Here are some examples.

```
mamba create --name test -c conda-forge python=3.11 openmm-ml
```

fails with the error

>   - package openmm-ml-1.0-pyhd8ed1ab_0 requires torchani >=2.2.2, but none of the providers can be installed

If I downgrade from python 3.11 to 3.10
```
mamba create --name test -c conda-forge python=3.10 openmm-ml
```

then it works and wants to install

```
   pytorch                   1.13.1  cuda112py310he33e0d6_200  conda-forge/linux-64     Cached
```

If I first create the environment with

```
mamba create --name test -c conda-forge python=3.10
```

then activate it and install openmm-ml with

```
mamba install -c conda-forge openmm-ml
```

then it installs a different, older pytorch version:

```
  pytorch                   1.12.1  cuda112py310he33e0d6_201  conda-forge/linux-64     488 MB
```

But if I try to use conda,

```
conda install -c conda-forge openmm-ml
```

then it fails with

```
ResolvePackageNotFound: 
  - python==3.10
```

**Comment by peastman** [comment]
I figured out how to reproduce the original problem: don't list conda-forge when creating the environment.

```
mamba create --name test python=3.10
conda activate test
mamba install -c conda-forge openmm-ml
```

That leads it to install

```
  pytorch                   1.12.1  cpu_py310h75c9ab6_0      conda-forge/linux-64     Cached
```

**Comment by sef43** [comment]
I can reproduce the problem using conda:
```
conda create --name test python=3.10
conda activate test
conda install -c conda-forge openmm-ml
```

mamba always works for me, either it gives a working env, or gives an error (torchani is not build for python 3.11).

I think once again the issue is conda struggling to deal with pytorch. The newer versions of pytorch seem to need a version of python that comes from the conda-forge channel rather than the main channel. By doing `conda create --name test python=3.10` it is installing python from pkgs/main channel, rather than conda-forge channel. If i use mamba it resinstalls python from conda-forge. Here are the examples.

**1. use conda and install python from main when creating the env:**
```
conda create -n test python=3.10
conda activate test
conda install -c conda-forge pytorch --dry-run
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/sfarr/miniconda3/envs/test

  added / updated specs:
    - pytorch


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cffi-1.15.0                |  py310h0fdd8cc_0         433 KB  conda-forge
    libblas-3.9.0              |   16_linux64_mkl          13 KB  conda-forge
    libcblas-3.9.0             |   16_linux64_mkl          12 KB  conda-forge
    liblapack-3.9.0            |   16_linux64_mkl          12 KB  conda-forge
    numpy-1.22.3               |  py310h4ef5377_2         6.8 MB  conda-forge
    pytorch-1.11.0             |cpu_py310h75c9ab6_1        47.2 MB  conda-forge
    ------------------------------------------------------------
                                           Total:        54.5 MB

The following NEW packages will be INSTALLED:

  cffi               conda-forge/linux-64::cffi-1.15.0-py310h0fdd8cc_0 
  intel-openmp       pkgs/main/linux-64::intel-openmp-2022.1.0-h9e868ea_3769 
  libblas            conda-forge/linux-64::libblas-3.9.0-16_linux64_mkl 
  libcblas           conda-forge/linux-64::libcblas-3.9.0-16_linux64_mkl 
  liblapack          conda-forge/linux-64::liblapack-3.9.0-16_linux64_mkl 
  libprotobuf        pkgs/main/linux-64::libprotobuf-3.20.1-h4ff587b_0 
  mkl                pkgs/main/linux-64::mkl-2022.1.0-hc2b9512_224 
  ninja              conda-forge/linux-64::ninja-1.11.0-h924138e_0 
  numpy              conda-forge/linux-64::numpy-1.22.3-py310h4ef5377_2 
  pycparser          conda-forge/noarch::pycparser-2.21-pyhd8ed1ab_0 
  python_abi         conda-forge/linux-64::python_abi-3.10-2_cp310 
  pytorch            conda-forge/linux-64::pytorch-1.11.0-cpu_py310h75c9ab6_1 
  sleef              conda-forge/linux-64::sleef-3.5.1-h9b69904_2 
  typing_extensions  conda-forge/noarch::typing_extensions-4.4.0-pyha770c72_0 

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2023.01.10~ --> conda-forge::ca-certificates-2022.12.7-ha878542_0 
  certifi            pkgs/main/linux-64::certifi-2022.12.7~ --> conda-forge/noarch::certifi-2022.12.7-pyhd8ed1ab_0 



DryRunExit: Dry run. Exiting.
```
conda installs `pytorch            conda-forge/linux-64::pytorch-1.11.0-cpu_py310h75c9ab6_1`

**2. use mamba and install python from main when creating the env:**
```
mamba create -n test python=3.10
mamba activate test
mamba install -c conda-forge pytorch --dry-run


Looking for: ['pytorch']

pkgs/main/linux-64                                            No change
pkgs/main/noarch                                              No change
pkgs/r/linux-64                                               No change
pkgs/r/noarch                                                 No change
conda-forge/noarch                                  11.1MB @   3.3MB/s  3.6s
conda-forge/linux-64                                29.4MB @   3.3MB/s  9.8s

Pinned packages:
  - python 3.10.*


Transaction

  Prefix: /home/sfarr/miniconda3/envs/test

  Updating specs:

   - pytorch
   - ca-certificates
   - certifi
   - openssl


  Package               Version  Build                     Channel                    Size
────────────────────────────────────────────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────────────────────────────────────────────

  + cffi                 1.15.1  py310h255011f_3           conda-forge/linux-64     Cached
  + cudatoolkit          11.8.0  h37601d7_11               conda-forge/linux-64     Cached
  + cudnn              8.4.1.50  hed8a83a_0                conda-forge/linux-64     Cached
  + icu                    70.1  h27087fc_0                conda-forge/linux-64     Cached
  + libblas               3.9.0  16_linux64_openblas       conda-forge/linux-64     Cached
  + libcblas              3.9.0  16_linux64_openblas       conda-forge/linux-64     Cached
  + libgfortran-ng       12.2.0  h69a702a_19               conda-forge/linux-64     Cached
  + libgfortran5         12.2.0  h337968e_19               conda-forge/linux-64     Cached
  + libhwloc              2.8.0  h32351e8_1                conda-forge/linux-64     Cached
  + libiconv               1.17  h166bdaf_0                conda-forge/linux-64     Cached
  + liblapack             3.9.0  16_linux64_openblas       conda-forge/linux-64     Cached
  + libnsl                2.0.0  h7f98852_0                conda-forge/linux-64     Cached
  + libopenblas          0.3.21  pthreads_h78a6416_3       conda-forge/linux-64     Cached
  + libprotobuf         3.21.12  h3eb15da_0                conda-forge/linux-64     Cached
  + libsqlite            3.40.0  h753d276_0                conda-forge/linux-64     Cached
  + libxml2              2.10.3  h7463322_0                conda-forge/linux-64     Cached
  + libzlib              1.2.13  h166bdaf_4                conda-forge/linux-64     Cached
  + llvm-openmp          15.0.7  h0cdce71_0                conda-forge/linux-64     Cached
  + magma                 2.6.2  hc72dce7_0                conda-forge/linux-64     Cached
  + mkl                2022.2.1  h84fe81f_16997            conda-forge/linux-64     Cached
  + nccl               2.14.3.1  h0800d71_0                conda-forge/linux-64     Cached
  + ninja                1.11.0  h924138e_0                conda-forge/linux-64     Cached
  + numpy                1.24.1  py310h8deb116_0           conda-forge/linux-64     Cached
  + pycparser              2.21  pyhd8ed1ab_0              conda-forge/noarch       Cached
  + python_abi             3.10  3_cp310                   conda-forge/linux-64     Cached
  + pytorch              1.13.1  cuda112py310he33e0d6_200  conda-forge/linux-64     Cached
  + sleef                 3.5.1  h9b69904_2                conda-forge/linux-64     Cached
  + tbb                2021.7.0  h924138e_1                conda-forge/linux-64     Cached
  + typing_extensions     4.4.0  pyha770c72_0              conda-forge/noarch       Cached

  Change:
────────────────────────────────────────────────────────────────────────────────────────────

  - _libgcc_mutex           0.1  main                      pkgs/main                      
  + _libgcc_mutex           0.1  conda_forge               conda-forge/linux-64     Cached
  - zlib                 1.2.13  h5eee18b_0                pkgs/main                      
  + zlib                 1.2.13  h166bdaf_4                conda-forge/linux-64     Cached

  Upgrade:
────────────────────────────────────────────────────────────────────────────────────────────

  - libgcc-ng            11.2.0  h1234567_1                pkgs/main                      
  + libgcc-ng            12.2.0  h65d4601_19               conda-forge/linux-64     Cached
  - libgomp              11.2.0  h1234567_1                pkgs/main                      
  + libgomp              12.2.0  h65d4601_19               conda-forge/linux-64     Cached
  - libstdcxx-ng         11.2.0  h1234567_1                pkgs/main                      
  + libstdcxx-ng         12.2.0  h46fd767_19               conda-forge/linux-64     Cached
  - libuuid              1.41.5  h5eee18b_0                pkgs/main                      
  + libuuid              2.32.1  h7f98852_1000             conda-forge/linux-64     Cached
  - openssl              1.1.1s  h7f8727e_0                pkgs/main                      
  + openssl               3.0.7  h0b41bf4_2                conda-forge/linux-64     Cached

  Downgrade:
────────────────────────────────────────────────────────────────────────────────────────────

  - _openmp_mutex           5.1  1_gnu                     pkgs/main                      
  + _openmp_mutex           4.5  2_kmp_llvm                conda-forge/linux-64     Cached
  - python               3.10.9  h7a1cb2a_0                pkgs/main                      
  + python               3.10.8  h4a9ceb5_0_cpython        conda-forge/linux-64     Cached

  Summary:

  Install: 29 packages
  Change: 2 packages
  Upgrade: 5 packages
  Downgrade: 2 packages

  Total download: 0 B

────────────────────────────────────────────────────────────────────────────────────────────


Dry run. Exiting.

DryRunExit: Dry run. Exiting.
```
mamba installs ` pytorch              1.13.1  cuda112py310he33e0d6_200  conda-forge/linux-64`
and changes python (and other packages) to conda-forge builds



**Solutions**
The solution as you have found is to make sure the python is a conda-forge build, there a several working ways to do this.

1. specify conda-forge when you create the environment with a specific python version
```
conda create -n openmm -c conda-forge python=3.10
conda activate openmm
conda install -c conda-forge openmm-ml
```

2. or make a completely empty environment first
```
conda create -n openmm 
conda activate openmm
conda install -c conda-forge openmm-ml
```
the newest compatible python will be installed from conda-forge

4. install everything when you create the environment
```
conda create -n openmm -c conda-forge openmm-ml
```
5. or faster with mamba
```
mamba create -n openmm -c conda-forge openmm-ml
```
will install the latest version of everything correctly.


The order you do things with conda matters, by giving it all requirements at the same time, in a completely clean environment, it can work them out correctly (although slowly).

**Comment by peastman** [comment]
Your option 4 doesn't work for me.  Installing everything with mamba still produces the error.

If someone is creating a completely new environment from scratch, there are ways to make this work.  But what if someone wants to install it into an existing environment that already includes other software they want to use along with OpenMM?  Is there a way to force it to install the correct pytorch?

**Comment by sef43** [comment]
> Your option 4 doesn't work for me. Installing everything with mamba still produces the error.

What mamba version do you have? I have version 1.2

> If someone is creating a completely new environment from scratch, there are ways to make this work. But what if someone wants to install it into an existing environment that already includes other software they want to use along with OpenMM? Is there a way to force it to install the correct pytorch?

On my machine with mamba 1.2 then it does correctly install pytorch in an existing environment, the only time it fails to install things are if you try and use python 3.11 where it complains that it cant install torchani

**Comment by sef43** [comment]
Using just conda if I use the flag `--override-channels` then it also installs correctly in an existing environment:
```
conda create -n test python=3.10
conda activate test
conda install -c conda-forge openmm-ml --override-channels
```

**Comment by peastman** [comment]
I posted the announcement on the forum.  I also created a PR to update the installation instruction in the OpenMM-ML README.

Hopefully the conda issues won't affect too many people.  If we start getting reports from people running into them, we may need to investigate the causes and see if there's a way to prevent them.

**Comment by peastman** [comment]
I was using mamba 0.10.0.  I just upgraded to 0.27.0, which is the newest version available on conda-forge.  It now installs a GPU pytorch rather than a CPU one.
